File: sample-chat-obj-c/sample-chat/Components/UsersInfoScreen/InfoUsersController/ViewedByViewController.m
//
//  ViewedByViewController.m
//  sample-chat
//
//  Created by Injoit on 07.03.2022.
//  Copyright © 2022 Quickblox. All rights reserved.
//

#import "ViewedByViewController.h"
#import "TitleView.h"
#import "Profile.h"
#import "ChatManager.h"
#import "QBUUser+Chat.h"
#import "UIViewController+Alert.h"
#import "Log.h"

NSString *const DELIVERED = @"Message delivered to";
NSString *const VIEWED = @"Message viewed by";

@interface ViewedByViewController () <QBChatDelegate>
@property (nonatomic, strong) QBChatMessage *message;
@property (strong, nonatomic) TitleView *navigationTitleView;
@property (nonatomic, strong) ChatManager *chatManager;
@end

@implementation ViewedByViewController
- (void)setMessageID:(NSString *)messageID {
    _messageID = messageID;
    [self setupMessage];
}

- (void)setMessage:(QBChatMessage *)message {
    _message = message;
    [self setupUsers];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.refreshControl = nil;
    [QBChat.instance addDelegate:self];
    self.chatManager = [ChatManager instance];
    
    UIBarButtonItem *backButtonItem = [[UIBarButtonItem alloc] initWithImage:[UIImage imageNamed:@"chevron"]
                                                                       style:UIBarButtonItemStylePlain
                                                                      target:self
                                                                      action:@selector(didTapBack:)];
    self.navigationItem.leftBarButtonItem = backButtonItem;
    backButtonItem.tintColor = UIColor.whiteColor;
    self.navigationTitleView = [[TitleView alloc] initWithFrame:CGRectZero];
    self.navigationItem.titleView = self.navigationTitleView;
    [self setupNavigationTitle];
}

- (void)viewWillAppear:(BOOL)animated {
    
    if (QBChat.instance.isConnected == NO) {
        [self showNoInternetAlertWithHandler:nil];
        return;
    }
}

#pragma mark - Setup
- (void)setupMessage {
    __weak __typeof(self)weakSelf = self;
    [ChatManager.instance messagesWithDialogID:self.dialogID extendedRequest:@{@"_id": self.messageID} skip:0 limit:1 success:^(NSArray<QBChatMessage *> * _Nonnull messages, Boolean isLast) {
        QBChatMessage *downloadedMessage = messages.firstObject;
        if (!downloadedMessage) {
            return;
        }
        weakSelf.message = downloadedMessage;
    } errorHandler:^(NSString * _Nullable error) {
        Log(@"[%@] setupMessage - Error: %@",  NSStringFromClass([ViewedByViewController class]), error);
    }];
}

- (void)setupNavigationTitle {
    NSString *title = self.action == ChatActionViewedBy ? VIEWED : DELIVERED;
    NSString *numberUsers = [NSString stringWithFormat:@"%@ members", @(self.userList.fetched.count)];
    [self.navigationTitleView setupTitleViewWithTitle:title subTitle:numberUsers];
}

- (void)setupUsers {
    NSPredicate *predicate = [NSPredicate predicateWithFormat:@"ID != %@", @(self.profile.ID)];
    if (self.action == ChatActionViewedBy) {
        NSArray<QBUUser *> *participants = [self.chatManager.storage usersWithIDs:self.message.readIDs];
        participants = [participants filteredArrayUsingPredicate:predicate];
        if (participants.count) {
            self.userList.fetched = participants.mutableCopy;
        }
    } else if (self.action == ChatActionDeliveredTo) {
        NSArray<QBUUser *> *participants = [self.chatManager.storage usersWithIDs:self.message.deliveredIDs];
        participants = [participants filteredArrayUsingPredicate:predicate];
        if (participants.count) {
            self.userList.fetched = participants.mutableCopy;
        }
    }
    [self.tableView reloadData];
    [self setupNavigationTitle];
}

- (void)tableView:(UITableView *)tableView
    configureCell:(UserTableViewCell *)cell
     forIndexPath:(NSIndexPath *)indexPath {
    cell.checkBoxView.hidden = YES;
    cell.checkBoxImageView.hidden = YES;
    cell.userInteractionEnabled = NO;
}

#pragma mark - Actions
- (void)didTapBack:(UIButton *)sender {
    [self.navigationController popViewControllerAnimated:YES];
}

#pragma mark - QBChatDelegate
- (void)chatDidReadMessageWithID:(NSString *)messageID dialogID:(NSString *)dialogID readerID:(NSUInteger)readerID {
    if (self.profile.ID == readerID
        || self.action != ChatActionViewedBy
        || ![self.message.ID isEqualToString:messageID]) {
        return;
    }
    QBChatMessage *readedMessage = [self.dataSource messageWithID:messageID];
    if (readedMessage) {
        NSMutableArray *readIDs = [readedMessage.readIDs mutableCopy];
        if ([readIDs containsObject:@(readerID)]) {
            return;
        }
        [readIDs addObject:@(readerID)];
        [readedMessage setReadIDs: [readIDs copy]];
        [self.dataSource updateMessage:readedMessage];
        
        self.message = readedMessage;
        [self setupUsers];
    }
}

- (void)chatDidDeliverMessageWithID:(NSString *)messageID dialogID:(NSString *)dialogID toUserID:(NSUInteger)userID {
    if (self.profile.ID == userID
        || self.action != ChatActionDeliveredTo
        || ![self.message.ID isEqualToString:messageID]) {
        return;
    }
    QBChatMessage *deliveredMessage = [self.dataSource messageWithID:messageID];
    if (deliveredMessage) {
        NSMutableArray *deliveredIDs = [deliveredMessage.deliveredIDs mutableCopy];
        if ([deliveredIDs containsObject:@(userID)]) {
            return;
        }
        [deliveredIDs addObject:@(userID)];
        [deliveredMessage setDeliveredIDs: [deliveredIDs copy]];
        [self.dataSource updateMessage:deliveredMessage];
        
        self.message = deliveredMessage;
        [self setupUsers];
    }
}

- (void)chatDidConnect {
    [self setupMessage];
}

- (void)chatDidReconnect {
    [self setupMessage];;
}

@end

File: sample-chat-swift/sample-chat-swift/Components/UsersInfoScreen/ViewedByViewController.swift
//
//  ViewedByViewController.swift
//  sample-chat-swift
//
//  Created by Injoit on 31.01.2022.
//  Copyright © 2022 quickBlox. All rights reserved.
//

import UIKit

struct UsersInfoConstant {
    static let delivered = "Message delivered to"
    static let viewed = "Message viewed by"
    static let noUser = "No user with that name"
}

class ViewedByViewController: UserListViewController {
    //MARK: - Properties
    var dialogID: String!
    var dataSource: ChatDataSource!
    var messageID: String! {
        didSet {
            setupMessage()
        }
    }
    
    private var message: QBChatMessage? {
        didSet {
            setupUsers()
        }
    }
    private var titleView = TitleView()
    private let chatManager = ChatManager.instance
    
    //MARK: - Life Cycle
    override func viewWillAppear(_ animated: Bool) {
        
        refreshControl = nil
        navigationItem.titleView = titleView
        setupNavigationTitleByAction()
        QBChat.instance.addDelegate(self)
        let backButtonItem = UIBarButtonItem(image: UIImage(named: "chevron"),
                                             style: .plain,
                                             target: self,
                                             action: #selector(didTapBack(_:)))
        navigationItem.leftBarButtonItem = backButtonItem
        backButtonItem.tintColor = .white
        if QBChat.instance.isConnected == false {
            showNoInternetAlert(handler: nil)
            return
        }
    }
    
    //MARK: - Actions
    @objc func didTapBack(_ sender: UIBarButtonItem) {
        QBChat.instance.removeDelegate(self)
        navigationController?.popViewController(animated: true)
    }
    
    //MARK: - Internal Methods
    private func setupMessage() {
        guard let ID = messageID else {
            return
        }
        chatManager.messages(withID: dialogID,
                             extendedRequest: ["_id": ID],
                             skip: 0,
                             limit: 1,
                             successCompletion: { [weak self] (messages, cancel) in
            guard let self = self, let message = messages.first else { return }
            self.message = message
        }, errorHandler: { (error) in
            debugPrint("[\(ViewedByViewController.className)] \(#function) Error: \(error?.localized ?? "Error")")
        })
    }
    
    private func setupNavigationTitleByAction() {
        let title = action == .viewedBy ? UsersInfoConstant.viewed : UsersInfoConstant.delivered
        let numberUsers = "\(self.userList.fetched.count) members"
        titleView.setupTitleView(title: title, subTitle: numberUsers)
    }
    
    private func setupUsers() {
        guard currentUser.isFull == true else {
            return
        }
        if let message = message, let action = action {
            if action == .viewedBy {
                //check and add users who read the message
                guard let readIDs = message.readIDs,
                      readIDs.isEmpty == false else {
                          return
                      }
                userList.fetched = chatManager.storage.users(withIDs: readIDs).filter({ $0.id != currentUser.ID })
            } else if action == .deliveredTo {
                //check and add users who delivered the message
                guard let deliveredIDs = message.deliveredIDs,
                      deliveredIDs.isEmpty == false else  {
                          return
                      }
                userList.fetched = chatManager.storage.users(withIDs: deliveredIDs).filter({ $0.id != currentUser.ID })
            }
        }
        setupNavigationTitleByAction()
        tableView.reloadData()
    }
    
    override func configure(_ cell: UserTableViewCell, for indexPath: IndexPath) {
        cell.checkBoxView.isHidden = true
        cell.checkBoxImageView.isHidden = true
        cell.isUserInteractionEnabled = false
    }
    
    //MARK: - QBChatDelegate
    func chatDidReadMessage(withID messageID: String, dialogID: String, readerID: UInt) {
        if Profile().ID == readerID
            || action != ChatAction.viewedBy
            || messageID != message?.id {
            return
        }
        guard let dataSource = dataSource,let message = self.message else {
            return
        }
        if message.readIDs?.contains(NSNumber(value: readerID)) == false {
            message.readIDs?.append(NSNumber(value: readerID))
            dataSource.updateMessage(message)
        }
        self.message = message
        setupUsers()
    }
    
    func chatDidDeliverMessage(withID messageID: String, dialogID: String, toUserID userID: UInt) {
        if Profile().ID == userID
            || action != ChatAction.deliveredTo
            || messageID != self.message?.id {
            return
        }
        guard let dataSource = dataSource,let message = self.message else {
            return
        }
        if message.deliveredIDs?.contains(NSNumber(value: userID)) == false {
            message.deliveredIDs?.append(NSNumber(value: userID))
            dataSource.updateMessage(message)
        }
        self.message = message
        setupUsers()
    }
    
    override func chatDidConnect() {
       setupMessage()
    }
    
    override func chatDidReconnect() {
        setupMessage()
    }
}
